text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,129 @@
+all regions. This can bring significant improvement for read tail latency.
+
+After the region becomes larger, the chance to hit 1PC becomes higher. And the RPC count in a
+transaction can also be reduced. However, the hotspot chance is also increased. For writes, we rely
+on PD to split the hotspot in time and increase concurrency. In addition, we can also use unorder
+apply to make single region apply logs faster. Since unorder apply is a standalone feature, I’m not
+going into details here.
+
+For read hotspots, split should be triggered by PD, which can utilize the global statics from all
+regions and nodes. For normal read requests, TiKV will need to split its range into smaller buckets
+according to statics to increase concurrency. When TiDB wants to do a scan, it sends the RPC once,
+and TiKV will split the requests into smaller concurrent jobs.
+
+In the past design, follower read has also been discussed to offload works from leader. But TiDB
+can’t predict the progress of a follower, so latency can also become unpredictable. It’s more
+controllable to let PD split hotspots and schedule to get balance.
+
+Now that a region can be in 10GiB, a full scan can make the response exceed the limit of gRPC,
+which is 4GiB. So instead of unary, we need to use server streaming RPC to return the response as
+stream. The stream can set a limit on memory or count to avoid too many pending responses
+exhausting memory.
+
+After a region becomes larger, it needs to be warmed up before serving requests. There are two
+common cases:
+
+- In the very beginning, writes will only happen in a single region, the problem can be solved by
+    presplit + scatter.
+- Because followers don’t serve read, so their data is usually not in cache. After transferring
+    the leader, serving a range scan can trigger at most 10GiB IO. To warm it up, we can extend
+    the existing pre transfer leader mechanism. Leader will send its bucket statics to follower
+    in pre transfer leader, follower can do a pre scan for hot buckets before accepting
+    transferring leader.
+
+### Flow report
+
+Buckets statistics are reported in a standalone stream. They will be reported more frequently
+than original region heartbeats, which is 60s. The new stream is expected to be reported every
+10 seconds when have changes.
+
+### Replication
+
+A large region can take minutes to be replicated, PD should make cost estimation on actual size
+instead of fixed value. To avoid logs being truncated during sending a snapshot, we should set
+the log gc limit based on the actual size too. Raft engine should be GA before dynamic size
+region, so we don't need to worry about the extra write amplifications.
+
+There are ways to optimize the replications, sending a sub-LSM tree for example. But these designs
+may be complicated. And we rely on further designs like separating LSM tree to optimize the cost
+for all.
```

Comment:
Ingesting a large snapshot may cause a large compaction, so that the SST file can stay at L0 too long time and slow down or stop write requests. To resolve the problem, we can split the large region into several SST files.

Reply:
Yes, I forget it. Now it's added.